### PR TITLE
Exclude generated and untestable files from codecov report

### LINKED ISSUE #3258: Exclude generated and untestable files from codecov

### PR DESCRIPTION
**Pull request by willie-yao**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR adds a script to the `make test-cover` target to remove any generated files or old API versions from the codecov report, resulting in a more accurate number being reported.

**Note:** This PR also fixes a reoccurring issue where the codecov test is being killed during the make lint step. This is caused by the most recent Go version, and the issue is fixed by setting the Go version for the GitHub action to 1.19.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3258

**Special notes for your reviewer**:
Codecov report before changes: https://app.codecov.io/github/kubernetes-sigs/cluster-api-provider-azure
Codecov report after changes: https://app.codecov.io/github/willie-yao/cluster-api-provider-azure/commit/903bf7a6b702f7207af8471fc34c98519f57f609/tree

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/codecov-ignore.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+echo "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1/zz_generated.deepcopy.go
+sigs.k8s.io/cluster-api-provider-azure/api/v1beta1/azurecluster_conversion.go
+sigs.k8s.io/cluster-api-provider-azure/api/v1beta1/azureclusteridentity_conversion.go
```
Comment (mboersma) [inline hack/codecov-ignore.sh:19]: I think this list is missing `azureclustertemplate_conversion.go`. 

Maybe we should create it dynamically?

```shell
find api/v1beta1 -regex '.*_conversion\.go' -exec echo sigs.k8s.io/cluster-api-provider-azure/{} \;
find exp/api/v1beta1 -regex '.*_conversion\.go' -exec echo sigs.k8s.io/cluster-api-provider-azure/{} \;
find exp/api/v1beta1 -regex '.*zz_generated.*\.go' -exec echo sigs.k8s.io/cluster-api-provider-azure/{} \;
```

Comment (willie-yao) [reply hack/codecov-ignore.sh]: Thanks for the suggestion!

File: .github/workflows/cover.yaml [hunk 2]
```diff
@@ -12,7 +12,7 @@ jobs:
     - uses: actions/checkout@v3
     - uses: actions/setup-go@v3
       with:
-        go-version: '^1.19'
+        go-version: '1.19'
```
Comment (willie-yao) [inline .github/workflows/cover.yaml:15]: Note to reviewers: Current codecov failures are caused by the most recent version of Go. This change fixes that.

https://github.com/kubernetes-sigs/cluster-api-provider-azure/actions/workflows/cover.yaml

File: hack/codecov-ignore.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+find exp/api/v1beta1 -regex '.*zz_generated.*\.go' -exec echo sigs.k8s.io/cluster-api-provider-azure/{} \; >> codecov-ignore.txt
+
+while read p || [ -n "$p" ] 
+do
+if [[ "${OSTYPE}" == "darwin"* ]]; then
+  sed -i '' "/${p//\//\\/}/d" ./coverage.out
+else
+  sed -i "/${p//\//\\/}/d" ./coverage.out
+fi
+done < ./codecov-ignore.txt
```
Comment (mboersma) [inline hack/codecov-ignore.sh:34]: Can you add a newline at the end? Seems to make most editors and environments happier.

**Review by mboersma — COMMENTED** [review]
This looks good. I'll take another look when the `shellcheck` complaints are addressed.

**Review by mboersma — COMMENTED** [review]
/lgtm

**Pull request merged**